### PR TITLE
add gstheme hooks

### PR DESCRIPTION
**Pull request by KyleCardoza**
This PR includes several very small changes to a few classes, all of which give GSTheme more ability to control the visual presentation of the UI. Each GSTheme hook is accompanied by a default implementation in GSTheme itself which simply does whatever the library was going to do already; the net effect is that these changes change nothing for existing apps and themes, but allow new themes more control.

### CODE REVIEW

File: Source/NSMenu.m
Comment (fredkiefer) [inline Source/NSMenu.m:285]: This can get rather confusing. now the theme may rearrange the menu and afterwords the code here will do something different with it. Wouldn't it be simpler to move all the code from here into the default implementation in the theme and allow other themes to override it?

File: Source/NSMenuItemCell.m
Comment (fredkiefer) [inline Source/NSMenuItemCell.m:294]: Again I am wondering whether we should move more of the original code into the default theme. As the code is now, the theme must be fully aware on how we build up the key here, parse that and rewrite it.
The only thing that speaks against that move is that this code here also changes the key equivalent font. So it is not side effect free.

File: Source/NSMenuView.m
Comment (fredkiefer) [inline Source/NSMenuView.m:792]: This should be an CGFloat and that type should also be used in your new method.

**Review by fredkiefer — CHANGES_REQUESTED** [review]
Looks good to me. There are a few places where more discussion is needed but the general direction is great.

File: Source/GSThemeDrawing.m
Comment (KyleCardoza) [reply Source/GSThemeDrawing.m:2087]: Right, my bad. Consider that one fixed.

File: Source/NSMenu.m
Comment (KyleCardoza) [reply Source/NSMenu.m:268]: Removed, must have been a slip of the hands.

File: Source/NSMenu.m
Comment (KyleCardoza) [reply Source/NSMenu.m:285]: You know, I was thinking about that -- there isn't any really good reason for that not to be the case. The whole point of GSTheme, as I understand it, is to allow overriding default behaviour without breaking existing code -- so by moving the default behaviour of things into the base GSTheme class, we would accomplish that end with more flexibility for theme authors and no cost to those who aren't interested in fancy new themes.

So anyway, I'll get that change done too.

File: Source/NSMenuItemCell.m
Comment (KyleCardoza) [reply Source/NSMenuItemCell.m:294]: I think you're right, as much as possible of the "default" drawing behaviour should be moved into the default GSTheme class, both for architectural reasons, and because it affords theme authors far 
more power and flexibility in customizing the visual presentation.

So perhaps the implementation of `_keyEquivalentFont` and `_setKeyEquivalentFont` should defer to the theme, as well as replacing my implementation of `GSTheme-[keyForKeyEquivalent]` with the existing body of `NSMenuItemCell-[_keyEquivalentString]`, and have that method itself defer to the matching theme method?

File: Source/NSMenuView.m
Comment (KyleCardoza) [reply Source/NSMenuView.m:792]: Got it, will fix.

**Comment by rmottola** [comment]
I have tested this with the standard theme and with the WinClassic theme which sets a horizontal menu and I see no issue.

**Comment by KyleCardoza** [comment]
Excellent to hear. Is there anything else I should work on before marking it ready for final review?

**Comment by rmottola** [comment]
> Excellent to hear. Is there anything else I should work on before marking it ready for final review?

that's up to @fredkiefer - I was just adding some additional tests & checks.

File: Source/GSThemeMenu.m
Comment (fredkiefer) [inline Source/GSThemeMenu.m:263]: `[GSTheme theme]` should just be `self` here.

File: Source/GSThemeDrawing.m
Comment (fredkiefer) [inline Source/GSThemeDrawing.m:2075]: Not your fault, but we shouldn't need this second method to determine the close button size.

File: Source/NSMenuItemCell.m
Comment (fredkiefer) [reply Source/NSMenuItemCell.m:294]: I fully agree with your comment, but your code does not reflect this.

**Review by fredkiefer — APPROVED** [review]

File: Source/NSMenuItemCell.m
Comment (KyleCardoza) [reply Source/NSMenuItemCell.m:294]: I was waiting for feedback on the idea before implementing it -- but I'm glad we're on the same page there. I will get that done, probably in my next PR, which I plan to be specifically about moving more default behaviour into the base GSTheme class, as opposed to the minimal set of hooks I needed for my own personal passion project.

**Comment by gcasa** [comment]
Hey @KyleCardoza if you deem this ready you are okay to merge it.   If you can't then either @fredkiefer or I can do it.  I may have to add you to the project to allow you to have permission.   I will do that now.

**Comment by KyleCardoza** [comment]
I didn't think I could do that, but if that's how we want to do it I am happy and proud to accept.

**Pull request merged**
